cpp/include/rapidsmpf/rmm_resource_adaptor.hpp

```diff
@@ -0,0 +1,236 @@
+     * This reflects the number of allocations that have not yet been deallocated.
+     *
+     * @param alloc_type The allocator type to query. Use `AllocType::ALL` to return the
+     * sum across all types.
+     * @return The number of active allocations for the specified type.
+     */
+    [[nodiscard]] std::uint64_t num_current_allocs(AllocType alloc_type = AllocType::ALL)
+        const noexcept;
+
+    /**
+     * @brief Returns the current memory usage in bytes for the specified allocator type.
+     *
+     * Current usage is the total bytes currently allocated but not yet deallocated.
+     *
+     * @param alloc_type The allocator type to query. Use `AllocType::ALL` to return the
+     * sum across all types.
+     * @return The current memory usage in bytes for the specified type.
+     */
+    [[nodiscard]] std::uint64_t current(AllocType alloc_type = AllocType::ALL)
+        const noexcept;
+
+    /**
+     * @brief Returns the total number of bytes allocated over the lifetime of this
+     * record.
+     *
+     * This value accumulates over time and is not reduced by deallocations.
+     *
+     * @param alloc_type The allocator type to query. Use `AllocType::ALL` to return
+     * the sum across all types.
+     * @return The total number of bytes allocated for the specified type.
+     */
+    [[nodiscard]] std::uint64_t total(AllocType alloc_type = AllocType::ALL)
+        const noexcept;
+
+    /**
+     * @brief Returns the peak memory usage (in bytes) for the specified allocator type.
+     *
+     * The peak represents the highest value reached by current memory usage over the
+     * lifetime of the allocator. It does not decrease after deallocations.
+     *
+     * When queried with `AllocType::ALL`, this returns the highest combined memory
+     * usage ever observed across both primary and fallback allocators, not the sum of
+     * individual peaks.
+     *
+     * @param alloc_type The allocator type to query. Defaults to `AllocType::ALL`.
+     * @return The peak memory usage in bytes for the specified type.
+     */
+    [[nodiscard]] std::uint64_t peak(AllocType alloc_type = AllocType::ALL)
+        const noexcept;
+
+    /**
+     * @brief Records a memory allocation event.
+     *
+     * Updates the allocation counters and memory usage statistics, and adjusts
+     * peak usage if the new current usage exceeds the previous peak.
+     *
+     * @param alloc_type The allocator that performed the allocation.
+     * @param nbytes     The number of bytes allocated.
+     */
+    void record_allocation(AllocType alloc_type, std::uint64_t nbytes);
+
+    /**
+     * @brief Records a memory deallocation event.
+     *
+     * Reduces the current memory usage for the specified allocator.
+     *
+     * @param alloc_type The allocator that performed the deallocation.
+     * @param nbytes     The number of bytes deallocated.
+     */
+    void record_deallocation(AllocType alloc_type, std::uint64_t nbytes);
```

Comment:
Maybe make those private and have `RmmResourceAdaptor` as `friend`? There's no mutex protecting these calls here, so arbitrarily calling them outside of `RmmResourceAdaptor` is most certainly unsafe in a multi-threaded context.

Reply:
+1 on this.

Reply:
I don't know, `ScopedMemoryRecord` is just a container of memory counters. I think it is fine if an user want to use `ScopedMemoryRecord` to track some memory allocations manually outside of `RmmResourceAdaptor`?

Reply:
> I don't know, `ScopedMemoryRecord` is just a container of memory counters. I think it is fine if an user want to use `ScopedMemoryRecord` to track some memory allocations manually outside of `RmmResourceAdaptor`?

Then what does that represent anyway? Likely it's being used for an external RMM resource, the memory counters will then just be garbage, both because the report is not going to be representative due to mixing internal and external numbers and because those methods don't ensure thread-safety given there's no internal mutex.

Reply:
I was thinking it could be used to track memory statistics from multiple ranks if we implement operator+. But generally, I don't think we need to impose restrictions just for the sake of it. Since it's just a POD, there's no need to complicate things. It's also nice that we can test it independently.

I have add some docs: a39cac75e0ee1429bb535a66f3707b8d29ed1e50

But if you insist, we can make them private :)

Reply:
I feel like `ScopeMemoryRecord` should be unique to the `RmmResourceAdaptor` and it would be best to make it not copy-able. For an example, 
```c++
RmmResourceAdapter mr ...
auto rec1 = mr.get_record();
mr.allocate(...);
auto rec2 = mr.get_record(); 
```
Now, `rec1 != rec2`, but I feel like this could be problematic. WHat if we only return a ref to the memory record and make it non-copyable?

And also, as @pentschev mentioned, since `record_(de)allocation` methods are public, users can arbitrarily call those methods, but it doesn't really do any (de)allocs on the resource.

Reply:
The point is that my premise assumes `ScopeMemoryRecord` should be a POD (i.e., trivially copyable), so it can be moved between nodes easily and without requiring any locking. However, if we want to support something like @nirandaperera's example, then `ScopeMemoryRecord` would need to include an internal lock. This would also require the Cython bindings to use `with nogil`, and make copying/serialization non-trivial.